master/internal/rm/kubernetesrm/jobs.go

```diff
@@ -2101,32 +2132,96 @@ func numSlots(slots model.SlotsSummary) int {
 func (j *jobsService) listJobsInAllNamespaces(
 	ctx context.Context, opts metaV1.ListOptions,
 ) ([]batchV1.Job, error) {
-	var res []batchV1.Job
-	for n, i := range j.jobInterfaces {
-		pods, err := i.List(ctx, opts)
-		if err != nil {
-			return nil, fmt.Errorf("error listing pods for namespace %s: %w", n, err)
-		}
+	allJobs, err := j.jobInterfaces[""].List(ctx, opts)
+	if k8error.IsForbidden(err) {
+		return j.listJobsInAllNamespacesFallback(ctx, opts)
+	} else if err != nil {
+		logrus.WithError(err).WithField("function", "listJobsInAllNamespaces").Error("error listing jobs in all namespace")
+		return nil, err
+	}
 
-		res = append(res, pods.Items...)
+	namespaces := set.FromKeys(j.jobInterfaces)
+	var jobsWeCareAbout []batchV1.Job
+	for _, j := range allJobs.Items {
+		if namespaces.Contains(j.Namespace) {
+			jobsWeCareAbout = append(jobsWeCareAbout, j)
+		}
 	}
+	return jobsWeCareAbout, nil
+}
 
+func (j *jobsService) listJobsInAllNamespacesFallback(
+	ctx context.Context,
+	opts metaV1.ListOptions,
+) ([]batchV1.Job, error) {
+	var g errgroup.Group
+	var res []batchV1.Job
+	var resLock sync.Mutex
+	for n, i := range j.jobInterfaces {
+		g.Go(func() error {
+			pods, err := i.List(ctx, opts)
+			if err != nil {
+				return fmt.Errorf("error listing pods for namespace %s: %w", n, err)
+			}
+			resLock.Lock()
+			res = append(res, pods.Items...)
+			resLock.Unlock()
+			return nil
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		return nil, err
+	}
 	return res, nil
 }
 
 func (j *jobsService) listPodsInAllNamespaces(
 	ctx context.Context, opts metaV1.ListOptions,
-) (*k8sV1.PodList, error) {
-	res := &k8sV1.PodList{}
-	for n, i := range j.podInterfaces {
-		pods, err := i.List(ctx, opts)
-		if err != nil {
-			return nil, fmt.Errorf("error listing pods for namespace %s: %w", n, err)
-		}
+) ([]k8sV1.Pod, error) {
+	allPods, err := j.podInterfaces[""].List(ctx, opts)
+	if k8error.IsForbidden(err) {
+		return j.listPodsInAllNamespacesFallback(ctx, opts)
+	} else if err != nil {
+		return nil, err
+	}
 
-		res.Items = append(res.Items, pods.Items...)
+	namespaces := set.FromKeys(j.podInterfaces)
+	var podsWeWant []k8sV1.Pod
+	for _, pod := range allPods.Items {
+		if namespaces.Contains(pod.Namespace) {
+			podsWeWant = append(podsWeWant, pod)
+		}
 	}
+	return podsWeWant, nil
+}
 
+func (j *jobsService) listPodsInAllNamespacesFallback(
+	ctx context.Context,
+	opts metaV1.ListOptions,
+) ([]k8sV1.Pod, error) {
+	var g errgroup.Group
```

Comment:
Is there a limit to the number of go funcs started at the same time? Or is it possible to wait on hundreds of go funcs?

Reply:
great question, no. i'd even added a `g.SetLimit` for this but the client they all use it already throttled so i didn't care very much to limit it.

Reply:
That makes sense!